firefox_puppeteer/api/software_update.py

```diff
@@ -0,0 +1,491 @@
+
+          return Services.urlFormatter.formatURL(arguments[0]);
+
+          """, script_args=[url])
+
+        if force:
+            if '?' in url:
+                url += '&'
+            else:
+                url += '?'
+            url += 'force=1'
+        return url
+
+
+class ActiveUpdate(BaseLib):
+
+    def __getattr__(self, attr):
+        value = self.marionette.execute_script("""
+          let ums = Cc['@mozilla.org/updates/update-manager;1']
+                    .getService(Ci.nsIUpdateManager);
+          return ums.activeUpdate[arguments[0]];
+          """, script_args=[attr])
+
+        if value:
+            return value
+        else:
+            raise AttributeError('{} has no attribute {}'.format(self.__class__.__name__,
+                                                                 attr))
+
+    @property
+    def exists(self):
+        """Is there an active update
+
+        :returns: True if there is an active update
+        """
+
+        active_update = self.marionette.execute_script("""
+          let ums = Cc['@mozilla.org/updates/update-manager;1']
+                    .getService(Ci.nsIUpdateManager);
+          return ums.activeUpdate;
+          """)
+
+        return bool(active_update)
+
+    def get_patch_at(self, patch_index):
+        """Use nsIUpdate.getPatchAt to return a patch from an update
+
+        :returns: JSON data for an nsIUpdatePatch object
+        """
+
+        return self.marionette.execute_script("""
+          let ums = Cc['@mozilla.org/updates/update-manager;1']
+                    .getService(Ci.nsIUpdateManager);
+          return ums.activeUpdate.getPatchAt(arguments[0]);
+          """, script_args=[patch_index])
+
+    @property
+    def patch_count(self):
+        """Get the patchCount from the active update
+
+        :returns: The patch count
+        """
+
+        return self.marionette.execute_script("""
+          let ums = Cc['@mozilla.org/updates/update-manager;1']
+                    .getService(Ci.nsIUpdateManager);
+          return ums.activeUpdate.patchCount;
+          """)
+
+    @property
+    def selected_patch(self):
+        """Get the selected patch for the active update
+
+        :returns: JSON data for the selected patch
+        """
+
+        return self.marionette.execute_script("""
+          let ums = Cc['@mozilla.org/updates/update-manager;1']
+                    .getService(Ci.nsIUpdateManager);
+          return ums.activeUpdate.selectedPatch;
+        """)
+
+
+class MARChannels(BaseLib):
+    """Class to handle the allowed MAR channels as listed in update-settings.ini
+    """
+
+    INI_SECTION = 'Settings'
+    INI_OPTION = 'ACCEPTED_MAR_CHANNEL_IDS'
+
+    def __init__(self, marionette_getter):
+        BaseLib.__init__(self, marionette_getter)
+
+        self.ini_file_path = self.marionette.execute_script("""
+          Cu.import('resource://gre/modules/Services.jsm');
+
+          let file = Services.dirsvc.get('GreD', Ci.nsIFile);
```

Comment:
I would advise to not use the shortcuts for `Components.*` here. It will fail if this code is executed from a window which does not have those defined. Please change that across the file.